src/ucp/tag/tag_send.c

```diff
@@ -16,35 +16,44 @@
 #include <ucs/debug/instrument.h>
 #include <string.h>
 
-static ucs_status_t ucp_tag_req_start_contig(ucp_request_t *req, size_t count,
-                                             ssize_t max_short, size_t zcopy_thresh,
-                                             size_t rndv_thresh,
-                                             const ucp_proto_t *proto)
+static ucs_status_t ucp_tag_req_start(ucp_request_t *req, size_t count,
+                                      ssize_t max_short, size_t zcopy_thresh,
+                                      size_t rndv_thresh,
+                                      const ucp_proto_t *proto)
 {
-    ucp_ep_config_t *config = ucp_ep_config(req->send.ep);
-    size_t only_hdr_size = proto->only_hdr_size;
+    ucp_ep_config_t *config  = ucp_ep_config(req->send.ep);
+    size_t only_hdr_size     = proto->only_hdr_size;
+    unsigned flag_iov_single = 1;
     ucs_status_t status;
-    size_t max_zcopy;
     ssize_t length;
 
-    length           = ucp_contig_dt_length(req->send.datatype, count);
+    if (UCP_DT_IS_IOV(req->send.datatype)) {
```

Comment:
need to avoid extra overhead for contig case. we move this code block after checking for `if ((length <= max_short) && !UCP_DT_IS_IOV(req->send.datatype)) {`, and use likely/unlikely macros